### ISSUE TITLE
Web - Login - Sign in with Google banner in ../sign-in-with-google page jumps

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#28080]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/sign-in-with-google
on Safari
2. Refresh the page

## Expected Result:
The Sign in with Google banner will not jump

## Actual Result:
The Sign in with Google banner jumps

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.73-0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/99e2f778-d813-4ec3-9979-42b1dc0524e3

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by maddylewis** [comment]
i can reproduce this on Safari: 

https://github.com/Expensify/App/assets/38016013/7ed9c30a-34f2-4c78-9fab-d726e79ef2b2

**Comment by MichaelAwinda** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Sign-in with Google banner keeps jumping up when page is refreshed
### What is the root cause of that problem?
`GoogleSigninButton` component is not rendered properly
### What changes do you think we should make in order to solve the problem?

Use `SocialIcon` component that has a built-in type for Google.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by solomon-ochepa** [comment]
Hi everyone, I'm Solomon Ochepa by name. I'm a full-stack developer (General PHP, Laravel, CodeIgniter, etc.) with more than ten (10) years experience.

I saw your post on the UpWork platform. Before proceeding, I'd like to know the stack.

Thanks.

**Comment by solomon-ochepa** [comment]
Contributor details
Expensify account email: solomonochepa@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01dfbdd135c64803fa

**Comment by studentofcoding** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

Web - Login - Sign in with Google banner in ../sign-in-with-google page jumps

## What is the root cause of that problem?

We call an external script on `src/components/SignInButtons/GoogleSignIn/index.website.js`, but the component is rendered first before all the styling finishes loading. This makes the layout shift, when the styling is loaded, and makes the button jump

https://github.com/Expensify/App/blob/3b2dc326a52127b6e5552b81c9f14e5e2c5a8126/src/components/SignInButtons/GoogleSignIn/index.website.js#L58-L69

If we inspect the network, we can see that the layout shift (button jump) is happening when font `4UabrENHsxJlGDuGo1OIlLU94YtzCwM.ttf` is loaded

<img width="1552" alt="Screenshot 2023-09-29 at 01 58 45" src="https://github.com/Expensify/App/assets/20473526/925bdef6-ee9e-4d9e-a8b3-87069b0ae320">


## What changes do you think we should make in order to solve the problem?

We need to wait for all the script and styling inside `GoogleSignIn` button to be finished first and then render the component after that.

This way when `GoogleSignIn` button shows, it won't jump.

## What alternative solutions did you explore? (Optional)

None

@maddylewis @situchan

**Comment by maddylewis** [comment]
@situchan - bump on proposal review - thanks!!

**Comment by situchan** [comment]
I don't see solid proposal yet

**Comment by studentofcoding** [comment]
> I don't see solid proposal yet

Do you have any feedback on my proposal?

We can see from the Network activity that the Button showed first and the layout shift (button jump) is happening when font 4UabrENHsxJlGDuGo1OIlLU94YtzCwM.ttf is downloaded & loaded as my proposal, thus I believe this is the Root Cause of it

**Comment by situchan** [comment]
> We need to wait for all the script and styling inside `GoogleSignIn` button to be finished first and then render the component after that.
>
> This way when `GoogleSignIn` button shows, it won't jump.

@studentofcoding I am not sure how you will implement this. Can you share demo video after applying solution?

**Comment by studentofcoding** [comment]
Sure @situchan here below are the result and code solution of it

## Result

https://github.com/Expensify/App/assets/20473526/91b7c84e-312b-4e08-9a65-3f9cec402b99




## Solution
```
React.useEffect(() => {
        const script = document.createElement('script');
        script.src = 'https://accounts.google.com/gsi/client';
        script.addEventListener('load', loadScript);
        script.async = true;
        script.onload = () => {
            loadScript();
            const googleButtonContainerDesktop = document.getElementById(desktopId);
            const googleButtonDekstop = googleButtonContainerDesktop.firstChild;

             // After the script is loaded, there is a styling change, therefore we intercept it by hide the button 1st and manually wait until all done (approx 750ms) show the button
            googleButtonDekstop.style.display = 'none';
            setTimeout(() => {
                googleButtonDekstop.style.display = 'block';
            }, 750);
        };
    
        document.body.appendChild(script);

        return () => {
            script.removeEventListener('load', loadScript);
            document.body.removeChild(script);
        };
    }, [loadScript]);
 ```

**Comment by studentofcoding** [comment]
Hey @situchan, I've just updated the video, as yesterday the link is broken, Please kindly check again [here](https://github.com/Expensify/App/issues/28080#issuecomment-1757032916) :)

**Comment by situchan** [comment]
@studentofcoding I checked your solution in detail but not deal to me

**Comment by studentofcoding** [comment]
Hey @situchan, Upon more dig and observation [here are the optimal solution](https://github.com/Expensify/App/issues/28080#issuecomment-1739872467)

TLDR: 
We need to track all the script loaded from google button via `window.performance.getEntriesByType('resource');` & use setTimout to wait for Font Change (from `https://fonts.gstatic.com/s/*`). 

This is happening because upon checking our entries tracker, the last logged activity is from `https://accounts.google.com/gsi/button` instead of `https://fonts.gstatic.com/s/*`, therefore we need additional setTimout to track for Font Change

**Comment by maddylewis** [comment]
@situchan - lmk your thoughts on this one https://github.com/Expensify/App/issues/28080#issuecomment-1768081034

**Comment by situchan** [comment]
> @situchan - lmk your thoughts on this one [#28080 (comment)](https://github.com/Expensify/App/issues/28080#issuecomment-1768081034)

I am sure that there's better & simple solution to fix jump issue.

Still awaiting proposals

**Comment by maddylewis** [comment]
@situchan - we really need an update! can you lmk your thoughts on this when you have a chance? ty! https://github.com/Expensify/App/issues/28080#issuecomment-1771294617

**Comment by situchan** [comment]
I already reviewed https://github.com/Expensify/App/issues/28080#issuecomment-1768081034. Still I think it's not ideal solution. We can just prevent this by fixing style.

**Comment by studentofcoding** [comment]
If the alternative is fixing it via changing the width and height for `googlePillButtonContainer`, like the one shared [here](https://stackoverflow.com/questions/69242615/prevent-sign-in-with-google-button-flickering-while-loading-when-centered-vertic) then it's only fixing the problem on Chromium-based Browser (Chrome, Brave, etc) and not Safari. 

Here are the demo via fixing style, below

- Chromium Browser

https://github.com/Expensify/App/assets/20473526/1adb7bda-f3e6-44b8-976c-e2b34663681b

- Safari

https://github.com/Expensify/App/assets/20473526/33de92d4-a17f-4456-bbd2-40cc44a5addc

Based on my last update, the RCA is not on the button styling (that we can control like above), but rather on how some Google Sign-In styles are loaded late. Thus, the only way to fix it is by showing the button after all the late styles are loaded.

cc: @situchan @maddylewis

**Comment by maddylewis** [comment]
i'll let @situchan confirm whether or not @studentofcoding's proposal makes sense to move forward with.

**Comment by situchan** [comment]
I am fine with @mkhutornyi's [proposal](https://github.com/Expensify/App/issues/28080#issuecomment-1784210296) as using same solution applied in circle button.
🎀 👀 🎀 C+ reviewed

**Comment by marcaaron** [comment]
This one landed on production a while ago. But seems like the issue got stuck. Did everyone who was supposed to get paid get paid here?

**Comment by situchan** [comment]
No payment yet

**Comment by maddylewis** [comment]
### Payment Summary
- [ ] Reporting bonus - N/A
- [ ] C+ - @situchan ($500 via Upwork)
- [ ] Contributor - @mkhutornyi ($500 via Upwork)

**Comment by maddylewis** [comment]
@mkhutornyi - remind me, are you a contractor or do you require payment through Upwork?

**Comment by maddylewis** [comment]
thanks! paid 👍
